csi/deployment_util.go

```diff
@@ -10,8 +10,110 @@ import (
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/pkg/util/pointer"
 )
 
+// CheckMountPropagationWithPodSpec https://github.com/kubernetes/kubernetes/issues/66086#issuecomment-404346854
+func CheckMountPropagationWithPodSpec(kubeClient *clientset.Clientset, image, namespace string) (bool, error) {
+	commonName := "longhorn-mount-propagation-tester"
+	mountName := "mountpoint"
+	mountPath := "/mnt/tmp"
+	podSpec := v1.PodSpec{
+		Containers: []v1.Container{
+			v1.Container{
+				Name:  commonName,
+				Image: image,
+				Args: []string{
+					"/bin/sh",
+					"-c",
+					"sleep infinity",
+				},
+				ImagePullPolicy: v1.PullIfNotPresent,
+				VolumeMounts: []v1.VolumeMount{
+					v1.VolumeMount{
+						Name:             mountName,
+						MountPath:        mountPath,
+						MountPropagation: &MountPropagationBidirectional,
+					},
+				},
+				SecurityContext: &v1.SecurityContext{
+					Privileged: pointer.BoolPtr(true),
+				},
+			},
+		},
+		Volumes: []v1.Volume{
+			v1.Volume{
+				Name: mountName,
+				VolumeSource: v1.VolumeSource{
+					HostPath: &v1.HostPathVolumeSource{
+						Path: mountPath,
+					},
+				},
+			},
+		},
+	}
+
+	daemonSet := &appsv1beta2.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      commonName,
+			Namespace: namespace,
+		},
+
+		Spec: appsv1beta2.DaemonSetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app": commonName,
+				},
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app": commonName,
+					},
+				},
+				Spec: podSpec,
+			},
+		},
+	}
+
+	daemonSetClient := kubeClient.AppsV1beta2().DaemonSets(namespace)
+	podClient := kubeClient.CoreV1().Pods(namespace)
+
+	logrus.Debugf("Trying to create the daemonset %s for MountPropagation check", commonName)
+	_, err := daemonSetClient.Create(daemonSet)
+	if err != nil {
+		return false, err
+	}
+	logrus.Debugf("Created the daemonset %s for MountPropagation check", commonName)
+
+	defer func() {
+		propagation := metav1.DeletePropagationForeground
+		if err := daemonSetClient.Delete(commonName,
+			&metav1.DeleteOptions{PropagationPolicy: &propagation}); err != nil {
+			logrus.Warnf("Failed to delete the daemonset %s for MountPropagation check", commonName)
+		}
+	}()
+
+	pods, err := podClient.List(metav1.ListOptions{LabelSelector: fmt.Sprintf("app=%s", commonName)})
```

Comment:
It may not be shown immediately in the pod list like the issue Jacie met before. Retry for a short period of time (e.g. 30s) if the pod doesn't show up.

Reply:
I have considered this problem, but cannot reproduce it on GKE.
It's ok, I added the retry logic.